<font color = 'purple'>**PROJECT DESCRIPTION**</font>

The coronavirus took the entire world by surprise, changing everyone's daily routine. City dwellers no longer spent their free time outside, going to cafes and malls; more people were home, reading books. That attracted the attention of startups that rushed to develop new apps for book lovers.
You've been given a database of one of the services competing in this market. It contains data on books, publishers, authors, and customer ratings and reviews of books. This information will be used to generate a value proposition for a new product

In [1]:
# import libraries
import pandas as pd
from sqlalchemy import create_engine


db_config = {'user': 'praktikum_student',         # user name
             'pwd': 'Sdf4$2;d-d30pp', # password
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432,              # connection port
             'db': 'data-analyst-final-project-db'}          # the name of the data base

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                                     db_config['pwd'],
                                                                       db_config['host'],
                                                                       db_config['port'],
                                                                       db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'})

In [2]:
books = pd.io.sql.read_sql('books', con = engine)
display(books.head())

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [3]:
books.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
book_id             1000 non-null int64
author_id           1000 non-null int64
title               1000 non-null object
num_pages           1000 non-null int64
publication_date    1000 non-null datetime64[ns]
publisher_id        1000 non-null int64
dtypes: datetime64[ns](1), int64(4), object(1)
memory usage: 47.0+ KB


In [4]:
authors = pd.io.sql.read_sql('authors', con = engine)
display(authors.head())

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [5]:
publishers = pd.io.sql.read_sql('publishers', con = engine)
display(publishers.head())

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [6]:
authors = pd.io.sql.read_sql('authors', con = engine)
display(authors.head())

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [33]:
ratings = pd.io.sql.read_sql('ratings', con = engine)
display(ratings.sample(10))

rating_id  book_id        username  rating
296         297       63        dwayne09       3
3181       3182      497    johngonzalez       5
3601       3602      583    ashleyjoshua       5
3778       3779      623        joseph14       3
2144       2145      350      jarvispaul       5
871         872      145         npowers       5
3354       3355      541        urussell       4
5906       5907      910     brentpowers       5
2390       2391      381     lesliegibbs       4
5883       5884      905  cassandragomez       5

In [8]:
ratings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
rating_id    6456 non-null int64
book_id      6456 non-null int64
username     6456 non-null object
rating       6456 non-null int64
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


In [9]:
reviews = pd.io.sql.read_sql('reviews', con = engine)
display(reviews.head())

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

- **Find the number of books released after January 1, 2000**

In [10]:
query = '''SELECT
COUNT(*) AS cnt,
COUNT(book_id) AS books_id_cnt,
COUNT(DISTINCT book_id) AS books_amount_uniq
FROM
books
WHERE
CAST(books.publication_date AS date) >= '2000-01-01'
'''

pd.io.sql.read_sql(query, con = engine)

cnt  books_id_cnt  books_amount_uniq
0  821           821                821

<font color = 'purple'>**The number of books released after January 1, 2000 is 821**</font>

- **Find the number of user reviews and the average rating for each book**

In [11]:
query = '''SELECT
COUNT(rating_id) AS review_amount,
COUNT(DISTINCT rating_id) review_amount_uniq
FROM
ratings
'''

pd.io.sql.read_sql(query, con = engine)

review_amount  review_amount_uniq
0           6456                6456

In [12]:
query = '''SELECT
books.book_id AS id,
books.title AS title,
COUNT(DISTINCT reviews.review_id) AS reviews_count,
AVG(ratings.rating) AS avg_rating
FROM books
RIGHT JOIN reviews ON reviews.book_id = books.book_id
RIGHT JOIN ratings ON ratings.book_id = books.book_id
GROUP BY
id, title
ORDER BY reviews_count DESC;'''

pd.io.sql.read_sql(query, con = engine)

id                                              title  reviews_count  \
0    948.0                            Twilight (Twilight  #1)              7   
1    854.0                                           The Road              6   
2    656.0                                     The Book Thief              6   
3    734.0                                   The Glass Castle              6   
4    963.0                                Water for Elephants              6   
..     ...                                                ...            ...   
990  465.0                  Naked Empire (Sword of Truth  #8)              1   
991  446.0                                Moo  Baa  La La La!              1   
992  431.0                Merrick (The Vampire Chronicles #7)              1   
993  196.0  Dragons of Winter Night (Dragonlance: Chronicl...              1   
994    NaN                                               None              0   

     avg_rating  
0      3.662500  
1      3.772727  
2      4.264151  
3      4.206897  
4      3.977273  
..          ...  
990    3.500000  
991    3.000000  
992    4.000000  
993    4.000000  
994    3.923077  

[995 rows x 4 columns]

<font color = 'purple'>**The top books by the average rating are Emily of New Moon and Stolen Hard Times and by number of user reviews Twilight  and The Road**</font>

- **Identify the publisher that has released the greatest number of books with more than 50 pages (this will help you exclude brochures and similar publications from your analysis)**

In [20]:
query = '''SELECT
publishers.publisher AS publisher,
COUNT(DISTINCT books.book_id) AS max_books
FROM books
INNER JOIN publishers ON publishers.publisher_id = books.publisher_id
WHERE
    num_pages >= 50
GROUP BY
publisher
ORDER BY max_books DESC;'''

pd.io.sql.read_sql(query, con = engine)

publisher  max_books
0                          Penguin Books         42
1                                Vintage         31
2               Grand Central Publishing         25
3                       Penguin Classics         24
4                                 Bantam         19
..                                   ...        ...
329                  Random House Anchor          1
330  Random House Audio Publishing Group          1
331  Random House Large Print Publishing          1
332                   Random House Trade          1
333                 Random House Vintage          1

[334 rows x 2 columns]

<font color = 'purple'>**The publisher that has released the greatest number of books with more than 50 pages is Penguin Books with 42 books**</font>

- **Identify the author with the highest average book rating: look only at books with at least 50 ratings**

In [47]:
query = '''SELECT
books.book_id AS books_id,
books.title AS title,
authors.author AS author,
AVG(book_rating) AS avg_book_rating

FROM 
books

LEFT JOIN authors ON books.author_id = authors.author_id

INNER JOIN( SELECT 
book_id, 
AVG(rating)
FROM 
ratings 
GROUP BY book_id
HAVING COUNT(rating)  >=  50) AS book_rating ON books.book_id = book_rating.book_id

GROUP BY
books_id, title, author, avg_book_rating
ORDER BY avg_book_rating DESC
'''

pd.io.sql.read_sql(query, con = engine)

books_id                                              title  \
0        302  Harry Potter and the Prisoner of Azkaban (Harr...   
1        722  The Fellowship of the Ring (The Lord of the Ri...   
2        299  Harry Potter and the Chamber of Secrets (Harry...   
3        656                                     The Book Thief   
4        300  Harry Potter and the Half-Blood Prince (Harry ...   
5        399                                       Little Women   
6        301  Harry Potter and the Order of the Phoenix (Har...   
7        750                The Hobbit  or There and Back Again   
8        779  The Lightning Thief (Percy Jackson and the Oly...   
9        405                                  Lord of the Flies   
10       696             The Da Vinci Code (Robert Langdon  #2)   
11       673                             The Catcher in the Rye   
12       627                                      The Alchemist   
13       545                                   Romeo and Juliet   
14       733                          The Giver (The Giver  #1)   
15        79                                        Animal Farm   
16        75               Angels & Demons (Robert Langdon  #1)   
17       948                            Twilight (Twilight  #1)   
18       488                                    Of Mice and Men   

                                               author  avg_book_rating  
0                          J.K. Rowling/Mary GrandPré         4.414634  
1                                      J.R.R. Tolkien         4.391892  
2                          J.K. Rowling/Mary GrandPré         4.287500  
3                   Markus Zusak/Cao Xuân Việt Khương         4.264151  
4                          J.K. Rowling/Mary GrandPré         4.246575  
5                                   Louisa May Alcott         4.192308  
6                          J.K. Rowling/Mary GrandPré         4.186667  
7                                      J.R.R. Tolkien         4.125000  
8                                        Rick Riordan         4.080645  
9                                     William Golding         3.901408  
10                                          Dan Brown         3.830508  
11                                      J.D. Salinger         3.825581  
12           Paulo Coelho/Alan R. Clarke/Özdemir İnce         3.789474  
13  William Shakespeare/Paul Werstine/Barbara A. M...         3.787879  
14                                         Lois Lowry         3.750000  
15            George Orwell/Boris Grabnar/Peter Škerl         3.729730  
16                                          Dan Brown         3.678571  
17                                    Stephenie Meyer         3.662500  
18                                     John Steinbeck         3.622951

<font color = 'purple'>**The author with the highest average book rating is J.K. Rowling/Mary GrandPré - Harry Potter and the Prisoner of Azkaban has the highest avg book rating 4.4**</font>

- **Find the average number of text reviews among users who rated more than 50 book**

In [15]:
book_id = f'''SELECT
AVG(SUBQUERY.n_reviews) AS avg_text_rev
FROM

   (SELECT 
    username,
    COUNT(DISTINCT review_id) AS n_reviews
    FROM reviews
    WHERE username IN
       (SELECT username AS username
        FROM ratings
        GROUP BY username
        HAVING COUNT(DISTINCT book_id) >50)
        GROUP BY username) AS SUBQUERY;'''



pd.io.sql.read_sql(book_id , con = engine)

avg_text_rev
0     24.333333

<font color = 'purple'>**The average number of text reviews among users who rated more than 50 book is 24**</font>